File: haskell/glue/src/Glue/Lib/Bool/Gt.hs
module Glue.Lib.Bool.Gt where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

gt :: IR Eval
gt = NativeFunc gtImpl

gtImpl :: IR Eval -> Eval (IR Eval)
gtImpl a = pure $ NativeFunc (gtRight a)

gtRight :: IR Eval -> IR Eval -> Eval (IR Eval)
gtRight a b = case (a, b) of
    (Integer na, Integer nb) -> pure . Bool $ na > nb
    (Float na, Float nb) -> pure . Bool $ na > nb
    (Integer na, Float nb) -> pure . Bool $ fromIntegral na > nb
    (Float na, Integer nb) -> pure . Bool $ na > fromIntegral nb
    _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/bool/gt.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Greater than comparison function
/// Mirrors Haskell Glue.Lib.Bool.Gt.gt exactly
final Ir gt = IrNativeFunc(gtImpl);

/// Greater than comparison implementation
/// Mirrors Haskell Glue.Lib.Bool.Gt.gtImpl exactly
Eval<Ir> gtImpl(Ir a) {
  return Eval.pure(IrNativeFunc(gtRight(a)));
}

/// Helper function for second argument
/// Mirrors Haskell Glue.Lib.Bool.Gt.gtRight exactly
Eval<Ir> Function(Ir) gtRight(Ir a) {
  return (Ir b) => switch ((a, b)) {
    (IrInteger(value: final na), IrInteger(value: final nb)) => Eval.pure(
      IrBool(na > nb),
    ),
    (IrFloat(value: final na), IrFloat(value: final nb)) => Eval.pure(
      IrBool(na > nb),
    ),
    (IrInteger(value: final na), IrFloat(value: final nb)) => Eval.pure(
      IrBool(na > nb),
    ),
    (IrFloat(value: final na), IrInteger(value: final nb)) => Eval.pure(
      IrBool(na > nb),
    ),
    _ => throwError(wrongArgumentType(['number', 'number'])),
  };
}
